source/common/network/lc_trie.h

```diff
@@ -0,0 +1,588 @@
+#pragma once
+
+#include <arpa/inet.h>
+
+#include <algorithm>
+#include <climits>
+#include <vector>
+
+#include "envoy/common/exception.h"
+#include "envoy/network/address.h"
+
+#include "common/common/assert.h"
+#include "common/common/empty_string.h"
+#include "common/network/cidr_range.h"
+#include "common/network/utility.h"
+
+#include "absl/numeric/int128.h"
+#include "fmt/format.h"
+
+namespace Envoy {
+namespace Network {
+namespace LcTrie {
+
+/**
+ * Level Compressed Trie for tagging IP addresses. Both IPv4 and IPv6 addresses are supported
+ * within this class with no calling pattern changes.
+ *
+ * The algorithm to build the LC-Trie is desribed in the paper 'IP-address lookup using LC-tries'
+ * by'S. Nilsson' and 'G. Karlsson'. Refer to LcTrieInternal for implementation and algorithm
+ * details.
+ */
+class LcTrie {
+public:
+  /**
+   * @param tag_data supplies a vector of tag and CIDR ranges.
+   * @param fill_factor supplies the fraction of completeness to use when calculating the branch
+   *                    value for a sub-trie.
+   * @param root_branching_factor supplies the branching factor at the root. The paper suggests for
+   *                              large LC-Tries to use the value '16' for the root branching
+   *                              factor. It reduces the depth of the trie.
+   */
+  LcTrie(const std::vector<std::pair<std::string, std::vector<Address::CidrRange>>>& tag_data,
+         double fill_factor = 0.5, uint32_t root_branching_factor = 0);
```

Comment:
Should we just make the default 16 for root_branching_factor?

Reply:
I think it may be better to use a much smaller default, like 6 or 8.  Or maybe auto-compute the root branching factor based on the number of prefixes submitted to the build function.

Here's my rationale...

With a root_branching factor of 16, each trie will use at least 256KB of memory: 2^16 nodes at the first level below the root, times 4 bytes per node.

When Nilsson _et al_ recommended 16 bits of branching at the root in their [original paper](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.108.5093&rep=rep1&type=pdf), their target use case was a router with at least tens of thousands of prefixes. In that scenario, dedicating 256KB of memory to shard the trie into 2^16 buckets seems like a reasonable tradeoff of memory for speed.

But proxy applications, if my own experience is representative, will usually have a much smaller number of prefixes configured: often dozens of prefixes, sometimes hundreds, rarely tens of thousands. For such configurations, 256KB strikes me as too much overhead, especially relative to the L1D and even L2 cache sizes of contemporary server CPUs.

Reply:
SGTM. Whatever you all think is best. Was mostly just wondering if the default should not be 0.

Reply:
I'll leave it at zero for now.